## Count or Frequency Encoding

In [3]:
import pandas as pd

In [48]:
train_set = pd.read_csv('http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data' , header = None,index_col=None)
train_set.head()

0                  1       2           3   4                    5   \
0  39          State-gov   77516   Bachelors  13        Never-married   
1  50   Self-emp-not-inc   83311   Bachelors  13   Married-civ-spouse   
2  38            Private  215646     HS-grad   9             Divorced   
3  53            Private  234721        11th   7   Married-civ-spouse   
4  28            Private  338409   Bachelors  13   Married-civ-spouse   

                   6               7       8        9     10  11  12  \
0        Adm-clerical   Not-in-family   White     Male  2174   0  40   
1     Exec-managerial         Husband   White     Male     0   0  13   
2   Handlers-cleaners   Not-in-family   White     Male     0   0  40   
3   Handlers-cleaners         Husband   Black     Male     0   0  40   
4      Prof-specialty            Wife   Black   Female     0   0  40   

               13      14  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3   United-States   <=50K  
4            Cuba   <=50K

In [49]:
columns = [1,3,5,6,7,8,9,13]

In [50]:
train_set = train_set[columns]

In [51]:
train_set.columns=['Employment','Degree','Status','Designation','family_job','Race','Sex','Country']

In [52]:
train_set.head()

Employment      Degree               Status         Designation  \
0          State-gov   Bachelors        Never-married        Adm-clerical   
1   Self-emp-not-inc   Bachelors   Married-civ-spouse     Exec-managerial   
2            Private     HS-grad             Divorced   Handlers-cleaners   
3            Private        11th   Married-civ-spouse   Handlers-cleaners   
4            Private   Bachelors   Married-civ-spouse      Prof-specialty   

       family_job    Race      Sex         Country  
0   Not-in-family   White     Male   United-States  
1         Husband   White     Male   United-States  
2   Not-in-family   White     Male   United-States  
3         Husband   Black     Male   United-States  
4            Wife   Black   Female            Cuba

In [53]:
### Find how many caregories are there in each feature

for feature in train_set.columns[:]:
    print(feature,":",len(train_set[feature].unique()),'labels')   

Employment : 9 labels
Degree : 16 labels
Status : 7 labels
Designation : 15 labels
family_job : 6 labels
Race : 5 labels
Sex : 2 labels
Country : 42 labels


In [ ]:
### here country has got the highest labels

In [54]:
train_set['Country'].value_counts().to_dict()

{' United-States': 29170,
 ' Mexico': 643,
 ' ?': 583,
 ' Philippines': 198,
 ' Germany': 137,
 ' Canada': 121,
 ' Puerto-Rico': 114,
 ' El-Salvador': 106,
 ' India': 100,
 ' Cuba': 95,
 ' England': 90,
 ' Jamaica': 81,
 ' South': 80,
 ' China': 75,
 ' Italy': 73,
 ' Dominican-Republic': 70,
 ' Vietnam': 67,
 ' Guatemala': 64,
 ' Japan': 62,
 ' Poland': 60,
 ' Columbia': 59,
 ' Taiwan': 51,
 ' Haiti': 44,
 ' Iran': 43,
 ' Portugal': 37,
 ' Nicaragua': 34,
 ' Peru': 31,
 ' France': 29,
 ' Greece': 29,
 ' Ecuador': 28,
 ' Ireland': 24,
 ' Hong': 20,
 ' Trinadad&Tobago': 19,
 ' Cambodia': 19,
 ' Laos': 18,
 ' Thailand': 18,
 ' Yugoslavia': 16,
 ' Outlying-US(Guam-USVI-etc)': 14,
 ' Hungary': 13,
 ' Honduras': 13,
 ' Scotland': 12,
 ' Holand-Netherlands': 1}

## Now we are going to replace the countries with the corrsponding number..for example: United stated with 29170, India with 100 and so on

### This process is called mapping

In [55]:
country_map = train_set['Country'].value_counts().to_dict()

In [56]:
train_set['Country'] = train_set['Country'].map(country_map)
train_set.head(10)

Employment      Degree                  Status         Designation  \
0          State-gov   Bachelors           Never-married        Adm-clerical   
1   Self-emp-not-inc   Bachelors      Married-civ-spouse     Exec-managerial   
2            Private     HS-grad                Divorced   Handlers-cleaners   
3            Private        11th      Married-civ-spouse   Handlers-cleaners   
4            Private   Bachelors      Married-civ-spouse      Prof-specialty   
5            Private     Masters      Married-civ-spouse     Exec-managerial   
6            Private         9th   Married-spouse-absent       Other-service   
7   Self-emp-not-inc     HS-grad      Married-civ-spouse     Exec-managerial   
8            Private     Masters           Never-married      Prof-specialty   
9            Private   Bachelors      Married-civ-spouse     Exec-managerial   

       family_job    Race      Sex  Country  
0   Not-in-family   White     Male    29170  
1         Husband   White     Male    29170  
2   Not-in-family   White     Male    29170  
3         Husband   Black     Male    29170  
4            Wife   Black   Female       95  
5            Wife   White   Female    29170  
6   Not-in-family   Black   Female       81  
7         Husband   White     Male    29170  
8   Not-in-family   White   Female    29170  
9         Husband   White     Male    29170

## Advantages
1. Easy to implement
2. Not increaing the feature space

## Disadvantages
1. It will provide the same weight if the frequencues are same